packages/rum/src/userActionCollection.ts

```diff
@@ -0,0 +1,160 @@
+import { generateUUID, Observable } from '@datadog/browser-core'
+import { getElementContent } from './getElementContent'
+import { LifeCycle, LifeCycleEventType, Subscription } from './lifeCycle'
+import { UserActionType } from './rum'
+
+const IDLE_DELAY = 100
+const BUSY_DELAY = 100
+const USER_ACTION_MAX_DURATION = 10_000
+
+interface UserActionEnded {
+  id: string
+  startTime: number
+  duration: number
+}
+
+function newUserAction(
+  changesObservable: Observable<ChangeEvent>,
+  finishCallback: (event: UserActionEnded | undefined) => void
+) {
+  if (currentUserAction) {
+    finishCallback(undefined)
+    return
+  }
+
+  let idleTimeoutId: ReturnType<typeof setTimeout>
+  const id = generateUUID()
+  const startTime = performance.now()
+  let hasFinished = false
+
+  const validationTimeoutId = setTimeout(() => finish(false), BUSY_DELAY)
+  const maxDurationTimeoutId = setTimeout(() => finish(true), USER_ACTION_MAX_DURATION)
+
+  currentUserAction = { id, startTime }
+
+  changesObservable.subscribe(({ isBusy }) => {
+    clearTimeout(validationTimeoutId)
+    clearTimeout(idleTimeoutId)
+    const time = performance.now()
+    if (!isBusy) {
+      idleTimeoutId = setTimeout(() => finish(true, time), IDLE_DELAY)
+    }
+  })
+
+  function finish(end: boolean, time = performance.now()) {
```

Comment:
Making two functions would not be ideal because they would share most of the same code. Instead, I changed the `finish` function like this: `finish(details: UserActionDetails | undefined)` to call the `finishCallback` with it. The logic is a bit more straightforward.